### ISSUE TITLE
Documentation for ConfusionMatrix is thin on details, and error messages not informative

### ISSUE THREAD

**Issue by nsinenian** [pytorch/ignite#1380]
## 🐛 Bug description

Confusion matrix requires that `y_pred` and `y` are to have different dimensions `(batch,num_classes)` and `(batch,)`, respectively. Getting this wrong results in the following error message, which makes no sense (batch size is 5 here, and both `y_pred` and `y` have shape `(5,1)`:

`y_pred must have shape (batch_size, num_categories, ...) and y must have shape of (batch_size, ...), but given torch.Size([5, 1]) vs torch.Size([5, 1])`

I believe that the `torch.Size([5, 1]) vs torch.Size([5, 1])` message should probably read `torch.Size([5, 1]) vs torch.Size([5,])` to illustrate the problem.

Separately, it would be good to mention in the `ConfusionMatrix` docs that one-hot encoded vectors are required, and that this can be implemented using `from ignite.utils.to_onehot` (I stumbled upon it after realizing I had to one-hot encode, since most people use `BCEWithLogitsLoss()` anyway, and hence their model predictions are not one-hot encoded).

## Environment

 - PyTorch Version (e.g., 1.4): 1.4.0
 - Ignite Version (e.g., 0.3.0): 0.4.2
 - OS (e.g., Linux): MacOS
 - How you installed Ignite (`conda`, `pip`, source): `pip`
 - Python version: 3.8
 - Any other relevant information:

**Comment by vfdev-5** [comment]
Thanks for the report @nsinenian ! 
Context : https://discuss.pytorch.org/t/ignite-confusionmatrix-metric-error/99079

**Comment by trsvchn** [comment]
> Documentation for ConfusionMatrix is thin on details, and error messages not informative

That's why we call it **Confusion matrix** :rofl:

**Comment by Afzal-Ind** [comment]
@vfdev-5 Hi, Could you take me through this code? Any link? Thanks.

**Comment by vfdev-5** [comment]
@Afzal-Ind please take a look here: https://github.com/pytorch/ignite/blob/master/ignite/metrics/confusion_matrix.py

**Comment by vfdev-5** [comment]
> Separately, it would be good to mention in the ConfusionMatrix docs that one-hot encoded vectors are required, and that this can be implemented using from ignite.utils.to_onehot (I stumbled upon it after realizing I had to one-hot encode, since most people use BCEWithLogitsLoss() anyway, and hence their model predictions are not one-hot encoded).

@nsinenian actually, this is not exactly true. Predictions are required to have shape `(batch_size, num_categories, ...)` (logits or after softmax), one-hot encoded. And i'd say, this something we directly have as network output for classification/segmentation tasks. As for targets, they should not be one-hot encoded: `(batch_size, ...)`
https://github.com/pytorch/ignite/blob/dca273fc73cdeefe0bf6858de64c69f23cbffe8b/ignite/metrics/confusion_matrix.py#L17-L19

**Comment by nsinenian** [comment]
@vfdev-5 The comment was with respect to predictions, not targets. For the predictions, I have (relevant excerpt):
````
__NUM_CLASSES__ = 2

def binary_one_hot_output_transform(output):
    """ Ouptuts a one-hot encoded vector with 0 or 1 values """
    [y_pred,y] = output
    [y_pred,y] = [torch.sigmoid(y_pred).round().long(),y]    
    return (to_onehot(y_pred,__NUM_CLASSES__), y.long())

valuation_metrics = {
        'confusion_matrix' : ConfusionMatrix(num_classes=__NUM_CLASSES__,output_transform=binary_one_hot_output_transform),
}

evaluator = create_supervised_evaluator(model,metrics=valuation_metrics,output_transform=lambda x,y,y_pred: (y_pred,y))
trainer = create_supervised_trainer(model,optimizer,criterion)

````
My model output (forward function) is:
````
    def forward(self,features):
      
        y = self.fc1(features)
        y = self.fcd(y)
        y = self.fc2(F.relu(y))

        return y
````

For binary classification, I use the `BCEWithLogitsLoss()` loss function which does not use one-hot encoded vectors (and hence the model output being what it is above). Perhaps my use case is atypical.

**Comment by vfdev-5** [comment]
@nsinenian thanks for the details. Yes, that's true that you have binary classification use-case and yes, it definitely makes sense to mention that how to deal with that in the docs !

**Comment by jrieke** [comment]
As far as I can see, the issue is only with binary classification, right?

I'd suggest changing the [docstring](https://pytorch.org/ignite/_modules/ignite/metrics/confusion_matrix.html) from

    - `y_pred` must contain logits and has the following shape (batch_size, num_categories, ...).

to

    - `y_pred` must contain logits and has the following shape (batch_size, num_categories, ...). If you are doing binary classification, you may first need to one-hot encode the network output, e.g. with `ignite.utils.to_onehot(torch.sigmoid(y_pred).round().long(), 2)`.

@vfdev-5 Ok for that?

**Comment by vfdev-5** [comment]
@jrieke I'd say the following have to be done here:
- add a note to `num_classes` arg that it should be > 1
- add this check into the code and raise ValueError otherwise
- add a basic test for that
- analyze the code and see if we still can have a message like 
```
y_pred must have shape (batch_size, num_categories, ...) and y must have shape of (batch_size, ...), but given torch.Size([5, 1]) vs torch.Size([5, 1])
```
if yes, make the message more user-friendly.
- add a more explicit note for the binary case. Maybe, a bit more polished version of this code snippet: https://github.com/pytorch/ignite/issues/1380#issuecomment-727253899 

What do you think ?

**Comment by jrieke** [comment]
@vfdev-5 Looks good! I implemented everything in #1607, feel free to make any more changes.

**Comment by vfdev-5** [comment]
@nsinenian we updated the docs in #1607, please take a look and feel free to comment out if something is not clear. Thanks !
